Initial Proposal for GitClient integration

zatimko opened the pull request:
### What does this PR do?

This is the initial integration proposal for a Git Client to test repo URLs and provide
both a Service Check and also a Latency metric on the Git Client's connectivity to the repo URL

### Motivation

There is not a way in Datadog today to test that a Git client can connect to a destination repo.

### Review checklist

- [X] PR has a meaningful title
- [X] Feature or bugfix has tests (Unit and Integration)
- [X] Git history is clean
- [X] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

I was unable to get mypy to pass successfully because there was no mypy.ini file in the root directory.  I added a mypy.ini file but I'm not sure why this would be missing if mypy tests are being executed. 🤔 

Also, I opened issue 322280 in Chat because I haven't been able to test the .whl successfully on my local system.  The `datadog-agent check` returns expected data, but Datadog UI does not show the Service Check or the Metric.

```
$ sudo datadog-agent check gitclient

=== Series ===
{
  "series": [
    {
      "metric": "gitclient.latency",
      "points": [
        [
          1584550768,
          673
        ]
      ],
      "tags": [],
      "host": "17105A-atimko",
      "type": "gauge",
      "interval": 0,
      "source_type_name": "System"
    }
  ]
}
=== Service Checks ===
[
  {
    "check": "gitclient",
    "host_name": "17105A-atimko",
    "timestamp": 1584550768,
    "status": 0,
    "message": "",
    "tags": []
  }
]
=========
Collector
=========

  Running Checks
  ==============

    gitclient (0.0.3)
    -----------------
      Instance ID: gitclient:7c11c0821c62899b [OK]
      Configuration Source: file:/opt/datadog-agent/etc/conf.d/gitclient.d/conf.yaml
      Total Runs: 1
      Metric Samples: Last Run: 1, Total: 1
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 1
      Average Execution Time : 685ms
```

florimondmanca commented:
Hi @zatimko! Thanks for this. Haven't looked at this yet, but…

> I was unable to get mypy to pass successfully because there was no mypy.ini file in the root directory. I added a mypy.ini file but I'm not sure why this would be missing if mypy tests are being executed. 🤔

This made us realize mypy hadn't been configured on this repo yet. Just added it in #565, so I went ahead and updated this PR, cleaning up `.mypy_cache` along the way. :-)

zatimko commented:
> This made us realize mypy hadn't been configured on this repo yet. Just added it in #565, so I went ahead and updated this PR, cleaning up `.mypy_cache` along the way. :-)

Ahh!  That makes sense why there wasn't a .gitignore entry for .mypy_cache.  --- Thanks!

hithwen commented on gitclient/datadog_checks/gitclient/data/conf.yaml.example:
```diff
@@ -0,0 +1,57 @@
+#
+instances:
+
+    ## @param git_http_repo_url - string - required
```
These need to go in `gitclient/assets/configuration/spec.yaml` then you run `ddev validate config --sync gitclient` to generate this file. You can [check apache core integration](https://github.com/DataDog/integrations-core/blob/master/apache/assets/configuration/spec.yaml) for how to edit the spec file:

hithwen commented on gitclient/datadog_checks/gitclient/gitclient.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class GitclientCheck(AgentCheck):
+    def check(self, instance):
```
Please use `self.instance` instead of the passed in `instance` a its a deprecated parameter

hithwen commented on gitclient/datadog_checks/gitclient/gitclient.py:
```diff
@@ -0,0 +1,43 @@
+        git_http_repo_url = instance.get('git_http_repo_url')
+
+        # Ensure input parameters exist
+        if not git_http_repo_url:
```
Please check for `ConfigurationError`s in `__init__` rather than the check method

hithwen commented on gitclient/setup.py:
```diff
@@ -0,0 +1,51 @@
+    long_description = f.read()
+
+
+CHECKS_BASE_REQ = 'datadog-checks-base>=4.2.0'
```
```suggestion
CHECKS_BASE_REQ = 'datadog-checks-base>=11.0.0'
```

hithwen commented:
Hi @zatimko. As part of new integration PR you'll also need to add the integration and youreself to .github/CODEOWNERS file

apigirl commented on gitclient/README.md:
```diff
@@ -0,0 +1,57 @@
+
+## Overview
+
+This check monitors that a [Git client][1] can connect to a target http/https repo url.  The host running the Datadog agent must already have a git client installed and configured for authentication.
```
```suggestion
This check monitors the connection between a [Git client][1] and a target HTTP/HTTPS repo URL. The host running the Datadog Agent must already have a git client installed and configured for authentication.
```

apigirl commented on gitclient/README.md:
```diff
@@ -0,0 +1,57 @@
+
+### Configuration
+
+1. Edit the `gitclient.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your gitclient performance data. See the [sample gitclient.d/conf.yaml][3] for all available configuration options.
```
```suggestion
1. Edit the `gitclient.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your GitClient performance data. See the [sample gitclient.d/conf.yaml][3] for all available configuration options.
```

apigirl commented on gitclient/README.md:
```diff
@@ -0,0 +1,57 @@
+
+### Metrics
+
+GitClient does not collect any [metrics][6].
```
```suggestion
See [metadata.csv][6] for a list of metrics provided by this integration.
```

apigirl commented on gitclient/README.md:
```diff
@@ -0,0 +1,57 @@
+
+### Metrics
+
+GitClient does not collect any [metrics][6].
```
yep. added a suggestion that adds the language needed for the documentation build to autogenerate this.

zatimko commented:
Thanks for all of the comments @hithwen and @apigirl ---

I'm currently stuck on a separate project getting OSX Chef cookbook/recipes for Datadog (this isn't supported yet in the current Chef Datadog cookbook), and that's a high priority action item for my team.

I'll return to this PR as soon as I can!
